### PR TITLE
gh-105481: add pseudo-instructions to the cases DSL

### PR DESCRIPTION
**Pull request by iritkatriel**
* Issue: gh-105481

### CODE REVIEW

File: Tools/cases_generator/generate_cases.py [hunk 1]
```diff
@@ -1044,6 +1048,13 @@ def write_metadata(self) -> None:
             self.out.emit("};")
             self.out.emit("#endif")
 
+    def write_pseudo_instrs(self) -> None:
+        """Write the IS_PSEUDO_OPCODE macro"""
```
Comment (iritkatriel) [inline Tools/cases_generator/generate_cases.py]:
```suggestion
        """Write the IS_PSEUDO_INSTR macro"""
```

**Comment by iritkatriel** [comment]
I'll look into adding unit tests for the generator.

**Comment by iritkatriel** [comment]
> I'll look into adding unit tests for the generator.

Actually the test are not running anymore, right? (I mean the ones in Tools/cases_generator/test_generator.py).

**Comment by gvanrossum** [comment]
> Actually the test are not running anymore, right? (I mean the ones in Tools/cases_generator/test_generator.py).

Sadly those must be run manually (since they use pytest) and I have neglected to do so, so they are broken. I've filed gh-105540 for that.

File: Python/bytecodes.c [hunk 2]
```diff
@@ -218,6 +219,10 @@ dummy_func(
             SETLOCAL(oparg, value);
         }
 
+        pseudo(STORE_FAST_MAYBE_NULL) = {
+            STORE_FAST
```
Comment (gvanrossum) [inline Python/bytecodes.c]: I'd suggest a trailing comma.

File: Tools/cases_generator/parser.py [hunk 3]
```diff
@@ -148,6 +153,8 @@ def definition(self) -> InstDef | Super | Macro | Family | None:
             return macro
         if family := self.family_def():
             return family
+        if pseudo := self.pseudo_def():
+            return pseudo
```
Comment (gvanrossum) [inline Tools/cases_generator/parser.py:157]: Please update the union in the return type. There are probably some other places to update as well.

File: Tools/cases_generator/generate_cases.py [hunk 4]
```diff
@@ -932,6 +957,19 @@ def effect_str(effects: list[StackEffect]) -> str:
                     self.error("Macro has virtual stack growth", thing)
                 popped = str(-low)
                 pushed = str(sp - low)
+            case parser.Pseudo():
+                instr = self.pseudos[thing.name]
+                pushed = popped = None
+                for target in self.pseudos[thing.name].targets:
+                    target_instr = self.instrs.get(target)
+                    assert target_instr
```
Comment (gvanrossum) [inline Tools/cases_generator/generate_cases.py:971]: Eventually some target might become a macro -- maybe at least add a comment?

File: Tools/cases_generator/generate_cases.py [hunk 5]
```diff
@@ -932,6 +957,19 @@ def effect_str(effects: list[StackEffect]) -> str:
                     self.error("Macro has virtual stack growth", thing)
                 popped = str(-low)
                 pushed = str(sp - low)
+            case parser.Pseudo():
+                instr = self.pseudos[thing.name]
+                pushed = popped = None
```
Comment (gvanrossum) [inline Tools/cases_generator/generate_cases.py]: I know it doesn't matter, but for consistency, I'd recommend always mentioning popped first. Also below in the condition on L968.

File: Tools/cases_generator/generate_cases.py [hunk 6]
```diff
@@ -932,6 +957,19 @@ def effect_str(effects: list[StackEffect]) -> str:
...
+                for target in self.pseudos[thing.name].targets:
```
Comment (gvanrossum) [inline Tools/cases_generator/generate_cases.py:966]: Add a comment here that this loop verifies all targets in a pseudo have the same stack effects.

FWIW maybe we also ought to check they have the same cache effect?

File: Tools/cases_generator/generate_cases.py [hunk 7]
```diff
@@ -932,6 +957,19 @@ def effect_str(effects: list[StackEffect]) -> str:
...
+                    target_instr = self.instrs.get(target)
+                    assert target_instr
+                    target_popped = effect_str(target_instr.input_effects)
+                    target_pushed = effect_str(target_instr.output_effects)
```
Comment (gvanrossum) [inline Tools/cases_generator/generate_cases.py:973]: You could probably just call `list_effect_size()`. That should also be the same.

File: Tools/cases_generator/generate_cases.py [hunk 8]
```diff
@@ -1021,9 +1069,9 @@ def write_metadata(self) -> None:
 
             # Write metadata array declaration
             self.out.emit("#ifndef NEED_OPCODE_METADATA")
-            self.out.emit("extern const struct opcode_metadata _PyOpcode_opcode_metadata[256];")
+            self.out.emit("extern const struct opcode_metadata _PyOpcode_opcode_metadata[512];")
```
Comment (gvanrossum) [inline Tools/cases_generator/generate_cases.py:1083]: Do we need an assert somewhere else ensuring that the highest opcode is < 512?

**Review by gvanrossum — COMMENTED** [review]
A great start. I suppose before you land this you are going to make it so that the new metadata is actually used? Or will that be a separate PR?

File: Tools/cases_generator/generate_cases.py (hunk 7 above)
Comment (iritkatriel) [reply Tools/cases_generator/generate_cases.py:973]: That would be fine for the loop, but at the end the function needs to return the effect_str.

Comment (gvanrossum) [reply Tools/cases_generator/generate_cases.py:973]: Oh. Of course. Never mind, then.

File: Tools/cases_generator/generate_cases.py (hunk 6 above)
Comment (iritkatriel) [reply Tools/cases_generator/generate_cases.py:966]: > maybe we also ought to check they have the same cache effect?

We check that the instr_fmts are the same. I think this includes the stack effects, no?

Comment (gvanrossum) [reply Tools/cases_generator/generate_cases.py:966]: Alas, no. The instruction format only cares about the bytecode sequence, not about stack effect.

Comment (gvanrossum) [reply Tools/cases_generator/generate_cases.py:966]: D'oh. Never mind. I read this backwards. You are 100% correct.

Comment (iritkatriel) [reply Tools/cases_generator/generate_cases.py:966]: Sorry, I meant cache effect. Were you referring to stack effect?

**Comment by gvanrossum** [comment]
I guess still waiting for the pseudo PR?

**Comment by iritkatriel** [comment]
I’m not sure where we would use the metadata for pseudo instructions. It probably makes sense to add assertions in the places where we map them to real instructions, that the metadata is the same. I’ll have a look tomorrow (or so).

**Review by gvanrossum — APPROVED** [review]
Sorry, I was confusing this PR with gh-105482. That one _uses_ the metadata. This one LGTM!

**Review by gvanrossum — APPROVED** [review]

**Comment by gvanrossum** [comment]
I found some bugs, here are fixes:
https://github.com/python/cpython/commit/4291adf7bf97b51cf1981cf1da053e583e0a293b

**Pull request merged**